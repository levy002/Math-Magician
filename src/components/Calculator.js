import React, { Component } from 'react';
import Button from './Button';

const operators = ['รท', 'x', '-', '+', '='];
const digits = ['AC', '+/-', '%', '7', '8', '9', '4', '5', '6', '1', '2', '3', '0', '.'];

class Calculator extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <ul className="calc-body">
        <li className="calc-screen">0</li>
        <li className="calc-keyboard">
          <Button buttons={digits} />
          <Button buttons={operators} />
        </li>
      </ul>
    );
  }
}

export default Calculator;
