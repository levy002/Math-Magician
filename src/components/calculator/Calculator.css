button {
  padding: 5%;
  font-size: 1.3rem;
  font-weight: bold;
  background-color: #fff;
  border: 1px solid rgb(241, 238, 238);
  cursor: pointer;
}

button:hover {
  background-color: rgba(255, 165, 0, 0.5);
}

.calc-body {
  display: flex;
  flex-flow: column;
  list-style-type: none;
  border: 2px solid rgb(241, 238, 238);
  width: 40%;
  padding: 0;
  margin: 150px auto 0;
}

.calc-screen {
  text-align: end;
  padding: 30px;
  background-color: rgb(241, 238, 238);
  font-weight: bold;
  font-size: 1.5rem;
}

.calc-keyboard {
  display: flex;
  width: 100%;
}

.calc-digits {
  width: 85%;
  display: grid;
  flex: 1;
  grid-template-columns: 1fr 1fr 1fr;
}

.calc-operators {
  width: 15%;
  display: flex;
  flex-flow: column;
}

.calc-operators button {
  flex: 1;
  margin: 1px;
  background-color: orange;
  border: 1px hidden;
}

.calc-operators button:hover {
  opacity: 0.5;
}

.calc-digits button:nth-last-child(2) {
  grid-column: span 2;
}
