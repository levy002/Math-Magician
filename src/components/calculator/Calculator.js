import React, { useState } from 'react';
import { MdOutlineDoubleArrow } from 'react-icons/md';
import calculate from '../logic/calculate';
import 'animate.css';
import './Calculator.css';

const operators = ['รท', 'x', '-', '+', '='];
const digits = [
  'AC',
  '+/-',
  '%',
  '7',
  '8',
  '9',
  '4',
  '5',
  '6',
  '1',
  '2',
  '3',
  '0',
  '00',
  '.',
];

const Calculator = () => {
  const [obj, calcObj] = useState({
    next: '',
    total: '',
    operation: '',
  });

  const handleClick = (btn) => {
    const { total, next, operation } = calculate(obj, btn);
    if (next === null && total === null) {
      calcObj({ next, total: '', operation });
    } else {
      calcObj({ next, total, operation });
    }
  };

  return (
    <div className="calc">
      <div className="calc-animation">
        <h3 className="animate__animated animate__pulse animate__infinite">Let&apos;s do some Maths</h3>
        <MdOutlineDoubleArrow className="arrow" />
      </div>
      <ul className="calc-body">
        <li data-testid="calc-screen" className="calc-screen">{obj.next || obj.total || '0'}</li>
        <li className="calc-keyboard">
          <div className="calc-digits">
            {digits.map((btn) => (
              <button
                onClick={() => handleClick(btn)}
                type="button"
                key={btn}
                id={btn}
              >
                {btn}
              </button>
            ))}
          </div>
          <div className="calc-operators">
            {operators.map((btn) => (
              <button onClick={() => handleClick(btn)} type="button" key={btn}>
                {btn}
              </button>
            ))}
          </div>
        </li>
      </ul>
    </div>
  );
};
export default Calculator;
