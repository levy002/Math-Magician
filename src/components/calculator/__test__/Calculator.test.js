import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Calculator from '../Calculator';

it('renders "AC" on Quotes link', () => {
  render(<Calculator />);
  const linkElement = screen.getByText("Let's do some Maths");
  expect(linkElement).toBeTruthy();
});

it('renders correctly', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('contains exactly 20 buttons', () => {
  render(<Calculator />);

  const buttons = screen.getAllByRole('button');

  expect(buttons.length).toBe(20);
});

it('returns the correct result after a {2, +, 2, =} operation', () => {
  render(<Calculator />);
  const calcscreen = screen.getByTestId(/calc-screen/);
  const twoButton = screen.getByText(/2/);
  const plusButton = screen.getByText(/^[+]$/);
  const equalButton = screen.getByText(/=/);

  fireEvent.click(twoButton);
  fireEvent.click(plusButton);
  fireEvent.click(twoButton);
  fireEvent.click(equalButton);

  expect(calcscreen.textContent).toBe('4');
});

it('returns the correct result after a {4, รท, 2, =} operation', () => {
  render(<Calculator />);
  const calcscreen = screen.getByTestId(/calc-screen/);
  const oneButton = screen.getByText(/4/);
  const plusButton = screen.getByText(/^[รท]$/);
  const twoButton = screen.getByText(/2/);
  const equalButton = screen.getByText(/=/);

  fireEvent.click(oneButton);
  fireEvent.click(plusButton);
  fireEvent.click(twoButton);
  fireEvent.click(equalButton);

  expect(calcscreen.textContent).toBe('2');
});
