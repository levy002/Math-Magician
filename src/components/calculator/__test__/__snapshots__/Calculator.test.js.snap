// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="calc"
>
  <div
    className="calc-animation"
  >
    <h3
      className="animate__animated animate__pulse animate__infinite"
    >
      Let's do some Maths
    </h3>
    <svg
      className="arrow"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M15.5 5H11l5 7-5 7h4.5l5-7z"
      />
      <path
        d="M8.5 5H4l5 7-5 7h4.5l5-7z"
      />
    </svg>
  </div>
  <ul
    className="calc-body"
  >
    <li
      className="calc-screen"
      data-testid="calc-screen"
    >
      0
    </li>
    <li
      className="calc-keyboard"
    >
      <div
        className="calc-digits"
      >
        <button
          id="AC"
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
        <button
          id="+/-"
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
        <button
          id="%"
          onClick={[Function]}
          type="button"
        >
          %
        </button>
        <button
          id="7"
          onClick={[Function]}
          type="button"
        >
          7
        </button>
        <button
          id="8"
          onClick={[Function]}
          type="button"
        >
          8
        </button>
        <button
          id="9"
          onClick={[Function]}
          type="button"
        >
          9
        </button>
        <button
          id="4"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
        <button
          id="5"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
        <button
          id="6"
          onClick={[Function]}
          type="button"
        >
          6
        </button>
        <button
          id="1"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
        <button
          id="2"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
        <button
          id="3"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
        <button
          id="0"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
        <button
          id="00"
          onClick={[Function]}
          type="button"
        >
          00
        </button>
        <button
          id="."
          onClick={[Function]}
          type="button"
        >
          .
        </button>
      </div>
      <div
        className="calc-operators"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          รท
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          x
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          -
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </div>
    </li>
  </ul>
</div>
`;
